### ISSUE TITLE
[BUG] Azure.Messaging.ServiceBus conflicts with Azure.Security.KeyVault.Secrets in Azure Functions

### ISSUE THREAD

**Issue by snapfisher** [Azure/azure-sdk-for-net#20864]
**Describe the bug**
Using Azure.Messaging.ServiceBus within an Azure Function (C#), causes Azure.Security.KeyVault.Secrets.SecretClient.GetSecret() to receive a 401 error on key vault access.  Commenting out the service bus pieces allows it all to work correctly.  Note that the service bus pieces do not use the Azure.Security.KeyVault.Secrets library at all.

**Expected behavior**
They should not interfere.  Making service bus calls should not affect key vault access

**Actual behavior (include Exception or Stack Trace)**
The actual get call to the key vault generates a 401 error.  This makes me think it's an authentication error, but the authentication is unchanged.  The key vault authentication is not used by the service bus SDK.

Note that is the service bus pieces are run in a separate function, then they work fine.

Just including the NuGet packages does not cause the error.

**To Reproduce**

```
DefaultAzureCredentialOptions credOPtions = new DefaultAzureCredentialOptions()
  {
      ExcludeEnvironmentCredential = true,
      ExcludeInteractiveBrowserCredential = true
  };

  var secretClient = new SecretClient(new Uri(configuration["kvUri"]), new DefaultAzureCredential(credOPtions));

  ImageData imageData = await RecognizeImage(filename, secretClient);

  recordId = await SaveToCosmosDB(configuration, imageData, secretClient);

 await SendSBMessageAsync(configuration["serviceBusConnectionString"], configuration["serviceBusName"], caption);

await ReceiveMessagesAsync(configuration["serviceBusConnectionString"], configuration["serviceBusName"]);


       private static async Task<string> SaveToCosmosDB(Microsoft.Extensions.Configuration.IConfigurationRoot Configuration,
            ImageData imageData, SecretClient secretClient)
        {
            string recordId = string.Empty;

            string dbEndpointUri = Configuration["dbEndpointUri"];
            string dbPrimaryKey = ((KeyVaultSecret)secretClient.GetSecret(Configuration["cosmosSecretName"])).Value;
        }
```

**Environment:**
Azure.Messaging.ServiceBus v7.1.2
Azure.Security.KeyVault.Secrets v4.1.1

Visual Studio Enterprise 16.9.4

PS C:\Users\pfisher\source\repos\forcemultiplier> dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.202
 Commit:    db7cc87d51

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.202\

Host (useful for support):
  Version: 5.0.5
  Commit:  2f740adc14

.NET SDKs installed:
  3.1.408 [C:\Program Files\dotnet\sdk]
  5.0.202 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team members best able to assist.

**Comment by snapfisher** [comment]
After looking at this for a real long time today.  I think part of the problem was that I did not realize that static variables are not reset between function calls, so I had a whole lot of clients just chugging away on this, simultaneously.  However, the 401 calls were read out of App Insights, so.....as screwed up as my code may be, I still don't see how this could have happened.

**Comment by heaths** [comment]
Was an exception being thrown with a 401? If you're looking at AppInsights, a 401 is expected as Key Vault issues a challenge. These will happen periodically when the challenge response token expires.

**Comment by snapfisher** [comment]
It’s hard to see how it could expire since I am creating a new credential object for each call.  And I’m only doing this once per function call.  These were failing in less than a minute.



Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by heaths** [comment]
A 401 is sent the first time, too - whenever a request must be authenticated a bearer token is required. If one is not available (as in the first request) or has expired, you will get a 401 response with the first request.

**Comment by snapfisher** [comment]
I get it, but if I just comment out the SB parts, the existing auth never fails.  The SB parts don’t use the key vault or the auth object, since they use the connection string which is an app setting in the function app configuration.  I can’t understand why, even if the SB code was completely incorrect, I would be getting key vault errors.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by snapfisher** [comment]
<img width="717" alt="kv401" src="https://user-images.githubusercontent.com/45605309/117345762-e963c000-ae74-11eb-9970-b301eb367f4e.PNG">

From AppInsights.  The code actually hung -- well it was not hung, but chugging along infinitely, so I did not get any stack trace.

I do understand what you are saying about a 401, but it is very strange.  I don't see this in app insights after I removed the SB parts.

**Comment by snapfisher** [comment]
If you think that the above is normal.  I'm happy to close it and move on.  There was still something funky going on, but its hard to say what.

**Comment by heaths** [comment]
Yes, that is normal. The first request without the `Authorization` header gets a 401 response. We save the original request and send an empty request to authenticate - in this case, to the MSI endpoint. After we get our bearer token challenge response, we send the original request back which you see as the 200 to the same secret endpoint (/secrets/cosmosdbkey/).

We'll add a FAQ for this. Was this the only sign of a seeming error you saw? If not client exception and the app worked correctly, feel free to close.
